python/mlx/nn/layers/recurrent.py

```diff
@@ -0,0 +1,287 @@
+        )
+
+    def __call__(self, x, hidden=None):
+        if self.b is not None:
+            x = mx.addmm(self.b, x, self.Wx)
+        else:
+            x = x @ self.Wx
+
+        x_rz = x[..., : -self.hidden_size]
+        x_n = x[..., -self.hidden_size :]
+
+        all_hidden = []
+
+        for idx in range(x.shape[-2]):
+            rz = x_rz[..., idx, :]
+            if hidden is not None:
+                h_proj = hidden @ self.Wh
+                h_proj_rz = h_proj[..., : -self.hidden_size]
+                h_proj_n = h_proj[..., -self.hidden_size :]
+
+                if self.bhn is not None:
+                    h_proj_n += self.bhn
+
+                rz = rz + h_proj_rz
+
+            rz = mx.sigmoid(rz)
+
+            r, z = mx.split(rz, 2, axis=-1)
+
+            n = x_n[..., idx, :]
+
+            if hidden is not None:
+                n = n + r * h_proj_n
+            n = mx.tanh(n)
+
+            hidden = (1 - z) * n
+            if hidden is not None:
+                hidden = hidden + z * hidden
+            all_hidden.append(hidden)
+
+        return mx.stack(all_hidden, axis=-2)
+
+
+class LSTM(Module):
+    r"""An LSTM recurrent layer.
+
+    The input has shape ``NLD`` or ``LD`` where:
+
+    * ``N`` is the optional batch dimension
+    * ``L`` is the sequence length
+    * ``D`` is the input's feature dimension
+
+    Concretely, for each element of the sequence, this layer computes:
+
+    .. math::
+        \begin{align*}
+        i_t &= \sigma (W_{xi}x_t + W_{hi}h_t + b_{i}) \\
+        f_t &= \sigma (W_{xf}x_t + W_{hf}h_t + b_{f}) \\
+        g_t &= \text{tanh} (W_{xg}x_t + W_{hg}h_t + b_{g}) \\
+        o_t &= \sigma (W_{xo}x_t + W_{ho}h_t + b_{o}) \\
+        c_{t + 1} &= f_t \odot c_t + i_t \odot g_t \\
+        h_{t + 1} &= o_t \text{tanh}(c_{t + 1})
+        \end{align*}
+
+    The hidden state :math:`h` and cell state :math:`c` have shape ``NH``
+    or ``H``, depending on whether the input is batched or not.
+
+    The layer returns two arrays, the hidden state and the cell state at
+    each time step, both of shape ``NLH`` or ``LH``.
+
+    Args:
+        input_size (int): Dimension of the input, ``D``.
+        hidden_size (int): Dimension of the hidden state, ``H``.
+        bias (bool): Whether to use biases or not. Default: ``True``.
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        bias: bool = True,
+    ):
+        super().__init__()
+
+        self.hidden_size = hidden_size
+        scale = 1.0 / math.sqrt(hidden_size)
+        self.Wx = mx.random.uniform(
+            low=-scale, high=scale, shape=(input_size, 4 * hidden_size)
+        )
+        self.Wh = mx.random.uniform(
+            low=-scale, high=scale, shape=(hidden_size, 4 * hidden_size)
+        )
+        self.bias = (
+            mx.random.uniform(low=-scale, high=scale, shape=(4 * hidden_size,))
+            if bias
+            else None
+        )
```

Comment:
We only want `self.bias` to be set if bias=True.

```suggestion
        if bias:
            self.bias = mx.random.uniform(low=-scale, high=scale, shape=(4 * hidden_size,))

Reply:
Yea that's more inline with how we do it in `nn.Linear`. In this case though we need the attribute set for later use. O/w we have to change how we check if the `bias` exists.

Reply:
Ok I see. I think it's better if we use the boolean `bias` to check the existence of a bias array. We would have to rename one of the `bias` attributes, but I think it's for the better because reusing the same variable with different types is confusing.

Another option is something like ```getattr(self, 'bias', None)``` to get an attribute and default to None when it doesn't exist.

This change would be nice to avoid an extra step when converting weights between different ml frameworks, but I'm also fine not doing it if I'm the only one that thinks so.

Reply:
> This change would be nice to avoid an extra step when converting weights between different ml frameworks

Could you expand on that? What does it make easier?

> Another option is something like getattr(self, 'bias', None) 

We can do `if "bias" in self` just like we do in the `Linear` layer. That is an easy change to make.

Reply:
You can ignore this. I thought that `self.bias = None` would complicated loading weights because it would **always** expect a `bias` array, but I realize now that `parameters()` only returns non null objects so it's not a problem.